File: sdk/kotlin/rocksky/src/test/kotlin/app/rocksky/RsqlTest.kt
package app.rocksky

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

class RsqlTest {
    @Test fun eqBareString() {
        assertEquals("artist==Radiohead", Filter.eq("artist", "Radiohead").build())
    }

    @Test fun eqQuotedString() {
        assertEquals("artist==\"Daft Punk\"", Filter.eq("artist", "Daft Punk").build())
    }

    @Test fun eqEscapesEmbeddedQuotes() {
        assertEquals("title==\"He said \\\"hi\\\"\"", Filter.eq("title", "He said \"hi\"").build())
    }

    @Test fun eqWildcardStaysBare() {
        assertEquals("artist==Daft*", Filter.eq("artist", "Daft*").build())
    }

    @Test fun ne() {
        assertEquals("artist!=Eminem", Filter.ne("artist", "Eminem").build())
    }

    @Test fun gt() {
        assertEquals("duration=gt=200000", Filter.gt("duration", 200000).build())
    }

    @Test fun ge() {
        assertEquals("year=ge=2000", Filter.ge("year", 2000).build())
    }

    @Test fun lt() {
        assertEquals("trackNumber=lt=5", Filter.lt("trackNumber", 5).build())
    }

    @Test fun le() {
        assertEquals("year=le=1999", Filter.le("year", 1999).build())
    }

    @Test fun isIn() {
        assertEquals("genre=in=(house,electro)", Filter.isIn("genre", listOf("house", "electro")).build())
    }

    @Test fun isOutQuotesUnsafeValues() {
        assertEquals("genre=out=(\"hip hop\")", Filter.isOut("genre", listOf("hip hop")).build())
    }

    @Test fun isNull() {
        assertEquals("uri==null", Filter.isNull("uri").build())
    }

    @Test fun isNotNull() {
        assertEquals("uri!=null", Filter.isNotNull("uri").build())
    }

    @Test fun andJoinsWithSemicolon() {
        val a = Filter.eq("artist", "Radiohead")
        val b = Filter.gt("duration", 200000)
        assertEquals("artist==Radiohead;duration=gt=200000", a.and(b).build())
    }

    @Test fun orJoinsWithComma() {
        val a = Filter.eq("artist", "Radiohead")
        val b = Filter.eq("artist", "Muse")
        assertEquals("artist==Radiohead,artist==Muse", a.or(b).build())
    }

    @Test fun andParenthesizesOrLeftOperand() {
        val a = Filter.eq("artist", "Radiohead")
        val b = Filter.eq("artist", "Muse")
        val c = Filter.gt("duration", 200000)
        assertEquals("(artist==Radiohead,artist==Muse);duration=gt=200000", a.or(b).and(c).build())
    }

    @Test fun andParenthesizesOrRightOperand() {
        val a = Filter.eq("artist", "Radiohead")
        val b = Filter.eq("genre", "house")
        val c = Filter.eq("genre", "electro")
        assertEquals("artist==Radiohead;(genre==house,genre==electro)", a.and(b.or(c)).build())
    }

    @Test fun orNeverParenthesizes() {
        val a = Filter.eq("artist", "Radiohead")
        val b = Filter.gt("duration", 200000)
        val c = Filter.eq("genre", "house")
        assertEquals("artist==Radiohead;duration=gt=200000,genre==house", a.and(b).or(c).build())
    }

    @Test fun dottedFieldSelector() {
        assertEquals("track.artist==\"Daft Punk\"", Filter.eq("track.artist", "Daft Punk").build())
    }

    @Test fun booleanRendersBare() {
        assertEquals("liked==true", Filter.eq("liked", true).build())
    }

    @Test fun isInEmptyListThrows() {
        assertFailsWith<IllegalArgumentException> { Filter.isIn("genre", emptyList()) }
    }

    @Test fun isOutEmptyListThrows() {
        assertFailsWith<IllegalArgumentException> { Filter.isOut("genre", emptyList()) }
    }

    @Test fun toStringEqualsBuild() {
        val f = Filter.eq("artist", "Daft Punk").and(Filter.gt("duration", 200000))
        assertEquals(f.build(), f.toString())
    }
}

File: sdk/clojure/test/rocksky/rsql_test.clj
(ns rocksky.rsql-test
  "Canonical RSQL builder vectors — identical across every Rocksky SDK.
  Pure: requires only rocksky.rsql, never the native-backed rocksky.core."
  (:require [clojure.test :refer [deftest is testing]]
            [rocksky.rsql :as rsql]))

(deftest comparisons
  (is (= "artist==Radiohead" (rsql/build (rsql/eq :artist "Radiohead"))))
  (is (= "artist==\"Daft Punk\"" (rsql/build (rsql/eq :artist "Daft Punk"))))
  (is (= "title==\"He said \\\"hi\\\"\""
         (rsql/build (rsql/eq :title "He said \"hi\""))))
  (is (= "artist==Daft*" (rsql/build (rsql/eq :artist "Daft*"))))
  (is (= "artist!=Eminem" (rsql/build (rsql/ne :artist "Eminem"))))
  (is (= "duration=gt=200000" (rsql/build (rsql/gt :duration 200000))))
  (is (= "year=ge=2000" (rsql/build (rsql/ge :year 2000))))
  (is (= "trackNumber=lt=5" (rsql/build (rsql/lt :trackNumber 5))))
  (is (= "year=le=1999" (rsql/build (rsql/le :year 1999))))
  (is (= "genre=in=(house,electro)"
         (rsql/build (rsql/in :genre ["house" "electro"]))))
  (is (= "genre=out=(\"hip hop\")" (rsql/build (rsql/out :genre ["hip hop"]))))
  (is (= "uri==null" (rsql/build (rsql/is-null :uri))))
  (is (= "uri!=null" (rsql/build (rsql/is-not-null :uri))))
  (is (= "liked==true" (rsql/build (rsql/eq :liked true)))))

(deftest combinators
  (let [a (rsql/eq :artist "Radiohead")]
    (is (= "artist==Radiohead;duration=gt=200000"
           (rsql/build (-> a (rsql/and (rsql/gt :duration 200000))))))
    (is (= "artist==Radiohead,artist==Muse"
           (rsql/build (-> a (rsql/or (rsql/eq :artist "Muse"))))))
    (testing "an :or operand is parenthesized inside AND"
      (is (= "(artist==Radiohead,artist==Muse);duration=gt=200000"
             (rsql/build (-> a
                             (rsql/or (rsql/eq :artist "Muse"))
                             (rsql/and (rsql/gt :duration 200000))))))
      (is (= "artist==Radiohead;(genre==house,genre==electro)"
             (rsql/build (-> a
                             (rsql/and (-> (rsql/eq :genre "house")
                                           (rsql/or (rsql/eq :genre "electro")))))))))
    (testing "OR never parenthesizes an AND operand"
      (is (= "artist==Radiohead;duration=gt=200000,genre==house"
             (rsql/build (-> a
                             (rsql/and (rsql/gt :duration 200000))
                             (rsql/or (rsql/eq :genre "house")))))))))

(deftest dotted-and-string-fields
  (testing ":track.artist reads as a plain keyword whose name keeps the dot"
    (is (nil? (namespace :track.artist)))
    (is (= "track.artist" (name :track.artist))))
  (is (= "track.artist==\"Daft Punk\""
         (rsql/build (rsql/eq :track.artist "Daft Punk"))))
  (testing "a namespaced keyword joins namespace and name with a dot"
    (is (= "track.artist==\"Daft Punk\""
           (rsql/build (rsql/eq :track/artist "Daft Punk")))))
  (testing "string fields are equivalent to keywords"
    (is (= (rsql/build (rsql/eq :artist "Radiohead"))
           (rsql/build (rsql/eq "artist" "Radiohead"))))
    (is (= (rsql/build (rsql/eq :track.artist "Daft Punk"))
           (rsql/build (rsql/eq "track.artist" "Daft Punk"))))))

(deftest empty-in-out-throws
  (is (thrown? clojure.lang.ExceptionInfo (rsql/in :genre [])))
  (is (thrown? clojure.lang.ExceptionInfo (rsql/out :genre []))))

(deftest build-accepts-strings
  (is (= "artist==Radiohead" (rsql/build "artist==Radiohead"))))
